robottelo/virtwho_utils.py

```diff
@@ -203,14 +204,28 @@ def get_configure_option(option, filename):
         raise VirtWhoError(f"option {option} is not exist or not be enabled in {filename}")
 
 
-def _get_hypervisor_mapping(logs, hypervisor_type):
+def get_rhsm_log():
+    """
+    Return the content of log file /var/log/rhsm/rhsm.log
+    """
+    _, logs = runcmd('cat /var/log/rhsm/rhsm.log')
+    return logs
+
+
+def _get_hypervisor_mapping(hypervisor_type):
     """Analysing rhsm.log and get to know: what is the hypervisor_name
     for the specific guest.
     :param str logs: the output of rhsm.log.
     :param str hypervisor_type: esx, libvirt, rhevm, xen, libvirt, kubevirt
     :raises: VirtWhoError: If hypervisor_name is None.
     :return: hypervisor_name and guest_name
     """
+    logs, _ = wait_for(
```

Comment:
This isn't quite right - `logs` is going to be empty, the label isn't assigned and wait_for doesn't return until the wait loop is complete.   `fail_condition` is not a function to be executed, the output of `func` is compared to `fail_condition` to evaluate whether or not to continue the loop.

Right now this is executing `get_rhsm_log`, which returns a string. `wait_for` is then checking `get_rhsm_log() == <fail_condition lambda>. 

You'll need to do something like this:

```
logs,  = wait_for(
    lambda: 'Host-to-guest mapping being sent to' in get_rhsm_log(),
    timeout=10,
    delay=2
```

That way, every iteration, it will call get_rhsm_log() and get a fresh copy of the log content, check whether the string is in it or not, resulting in a boolean. This boolean is then checked against fail_condition, which by default is `False`.  So if the string wasn't in the log, the func result is False, which is equal to the fail_condition, and the looping continues.

https://github.com/RedHatQE/wait_for/blob/master/wait_for/__init__.py#L146

Reply:
> This isn't quite right - `logs` is going to be empty, the label isn't assigned and wait_for doesn't return until the wait loop is complete. `fail_condition` is not a function to be executed, the output of `func` is compared to `fail_condition` to evaluate whether or not to continue the loop.
> 
> Right now this is executing `get_rhsm_log`, which returns a string. `wait_for` is then checking `get_rhsm_log() == <fail_condition lambda>.
> 
> You'll need to do something like this:
> 
> ```
> logs,  = wait_for(
>     lambda: 'Host-to-guest mapping being sent to' in get_rhsm_log(),
>     timeout=10,
>     delay=2
> ```
> 
> That way, every iteration, it will call get_rhsm_log() and get a fresh copy of the log content, check whether the string is in it or not, resulting in a boolean. This boolean is then checked against fail_condition, which by default is `False`. So if the string wasn't in the log, the func result is False, which is equal to the fail_condition, and the looping continues.
> 
> https://github.com/RedHatQE/wait_for/blob/master/wait_for/__init__.py#L146

@mshriver  when we use like this, the logs  value is True, in fact, we need to return the content of logs,  and we should analyze the log in the following code. 
In fact, whether the  'Host-to-guest mapping being sent to' in get_rhsm_log() or not , the content of logs from wait_for need to be returned to "logs" parameter.
Then we will analyze the logs in the following code.

Reply:
Thanks for making the update!  This is a consequence of the pattern in this case - you can't always get back the last value of the core function. There are tricky ways around it by having the `get_logs` function return a tuple, but its not really worth the complication. You did the right thing, just adding another call to `get_logs` after wait_for completes.